File: src/main/java/earth/terrarium/tempad/client/components/LocationPanel.java
package earth.terrarium.tempad.client.components;

import com.teamresourceful.resourcefullib.client.components.selection.SelectionList;
import earth.terrarium.tempad.common.Tempad;
import earth.terrarium.tempad.common.data.LocationData;
import net.minecraft.network.chat.Component;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.UUID;
import java.util.function.Consumer;
import java.util.function.Predicate;

public class LocationPanel extends SelectionList<TextEntry> {

    private static final List<TextEntry> EMPTY = List.of(
        new TextEntry(Component.translatable("gui." + Tempad.MODID + ".no_locations.first_line")),
        new TextEntry(Component.translatable("gui." + Tempad.MODID + ".no_locations.second_line"))
    );

    private final List<LocationData> locations;
    private final Predicate<UUID> isFavorite;

    public LocationPanel(
        int x, int y, int width, int height,
        List<LocationData> locations, Predicate<UUID> isFavorite,
        Consumer<@Nullable TextEntry> onSelection
    ) {
        super(x, y, width, height, 11, onSelection);
        this.locations = locations;
        this.isFavorite = isFavorite;
    }

    public void select(LocationData selected) {
        for (TextEntry entry : this.children()) {
            if (entry.data != null && entry.data.equals(selected)) {
                this.setSelected(entry);
                break;
            }
        }
    }

    public void update(String text) {
        if (this.locations.isEmpty()) {
            updateEntries(EMPTY);
        } else {
            updateEntries(this.locations.stream()
                .filter(data -> text.isBlank() || data.name().toLowerCase().contains(text.toLowerCase()))
                .map(locationData -> new TextEntry(locationData, locData -> this.isFavorite.test(locData.id())))
                .sorted(TextEntry::compareTo)
                .toList());
        }
    }
}

File: src/main/kotlin/earth/terrarium/tempad/client/widgets/LocationPanel.kt
package earth.terrarium.tempad.client.widgets

import com.teamresourceful.resourcefullib.client.components.selection.SelectionList
import earth.terrarium.tempad.Tempad
import earth.terrarium.tempad.api.locations.LocationData
import earth.terrarium.tempad.api.locations.ProviderSettings
import net.minecraft.network.chat.Component
import java.util.*

class LocationPanel(
    x: Int, y: Int, width: Int, height: Int,
    val locations: Map<ProviderSettings, Map<UUID, LocationData>>,
    private val isFavorite: (ProviderSettings?, UUID?) -> Boolean,
    onSelection: (TextEntry?) -> Unit
) : SelectionList<TextEntry>(x, y, width, height, 11, onSelection) {
    companion object {
        private val EMPTY: List<TextEntry> = listOf(
            TextEntry(Component.translatable("gui.${Tempad.MOD_ID}.no_locations.first_line")),
            TextEntry(Component.translatable("gui.${Tempad.MOD_ID}.no_locations.second_line")),
        )
    }

    init {
        for ((provider, locations) in locations) {
            addEntry(TextEntry(Component.translatable(provider.id.toLanguageKey("provider"))))
            for ((id, locationData) in locations) {
                val entry = TextEntry(provider, id, locationData, isFavorite)
                this.addEntry(entry)
            }
        }
    }

    fun select(settings: ProviderSettings, selected: UUID) {
        for (entry in this.children()) {
            if (entry.settings == settings && entry.id == selected) {
                this.setSelected(entry)
                return
            }
        }
    }

    fun update(text: String) {
        if (locations.isEmpty()) {
            updateEntries(EMPTY)
        } else {
            val newEntries = mutableListOf<TextEntry>()
            for ((provider, locations) in locations) {
                newEntries += (TextEntry(Component.translatable(provider.id.toLanguageKey("provider"))))
                for ((id, locationData) in locations) {
                    if (!locationData.name.contains(text, ignoreCase = true)) continue
                    newEntries += TextEntry(provider, id, locationData, isFavorite)
                }
            }
        }
    }
}